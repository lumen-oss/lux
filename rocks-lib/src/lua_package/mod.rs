use eyre::{eyre, Result};
use itertools::Itertools;
use mlua::FromLua;
use serde::{de, Deserialize, Deserializer, Serialize};
use std::{fmt::Display, str::FromStr};
pub use version::{PackageVersion, PackageVersionReq};

mod outdated;
mod version;

pub struct LuaPackage {
    name: PackageName,
    version: PackageVersion,
}

impl LuaPackage {
    pub fn new(name: PackageName, version: PackageVersion) -> Self {
        Self { name, version }
    }
    pub fn parse(name: String, version: String) -> Result<Self> {
        Ok(Self::new(
            PackageName::new(name),
            PackageVersion::parse(&version)?,
        ))
    }
    pub fn name(&self) -> &PackageName {
        &self.name
    }
    pub fn version(&self) -> &PackageVersion {
        &self.version
    }
    pub fn as_package_req(&self) -> LuaPackageReq {
        LuaPackageReq {
            name: self.name.clone(),
            version_req: self.version.clone().into(),
        }
    }
}

/// A lua package requirement with a name and an optional version requirement.
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "clap", derive(clap::Args))]
pub struct LuaPackageReq {
    /// The name of the package.
    name: PackageName,
    /// The version requirement, for example "1.0.0" or ">=1.0.0".
    #[cfg_attr(feature = "clap", clap(default_value_t = PackageVersionReq::default()))]
    version_req: PackageVersionReq,
}

impl LuaPackageReq {
    pub fn new(name: String, version: Option<String>) -> Result<Self> {
        Ok(Self {
            name: PackageName::new(name),
            version_req: match version {
                Some(version_req_str) => PackageVersionReq::parse(version_req_str.as_str())?,
                None => PackageVersionReq::default(),
            },
        })
    }
    pub fn parse(pkg_constraints: &String) -> Result<Self> {
        Self::from_str(&pkg_constraints.to_string())
    }
    pub fn name(&self) -> &PackageName {
        &self.name
    }
    pub fn version_req(&self) -> &PackageVersionReq {
        &self.version_req
    }
    /// Evaluate whether the given package satisfies the package requirement
    /// given by `self`.
    pub fn matches(&self, package: &LuaPackage) -> bool {
        self.name == package.name && self.version_req.matches(&package.version)
    }
}

impl FromStr for LuaPackageReq {
    type Err = eyre::Error;

    fn from_str(str: &str) -> Result<Self> {
        let rock_name_str = str
            .chars()
            .peeking_take_while(|t| t.is_alphanumeric() || matches!(t, '-' | '_' | '.'))
            .collect::<String>();

        if rock_name_str.is_empty() {
            return Err(eyre!(
                "Could not parse dependency name from {}",
                str.to_string()
            ));
        }

        let constraints = str.trim_start_matches(&rock_name_str).trim();
        let version_req = match constraints {
            "" => PackageVersionReq::default(),
            constraints => PackageVersionReq::parse(constraints.trim_start())?,
        };
        Ok(Self {
            name: PackageName::new(rock_name_str),
            version_req,
        })
    }
}

impl<'de> Deserialize<'de> for LuaPackageReq {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        Self::from_str(&s).map_err(de::Error::custom)
    }
}

/// A luarocks package name, which is always lowercase
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Hash)]
pub struct PackageName {
    name: String,
}

impl PackageName {
    pub fn new(name: String) -> Self {
        Self {
            // TODO: validations?
            name: name.to_lowercase(),
        }
    }
}

impl<'de> Deserialize<'de> for PackageName {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        Ok(PackageName::new(String::deserialize(deserializer)?))
    }
}

impl<'lua> FromLua<'lua> for PackageName {
    fn from_lua(
        value: mlua::prelude::LuaValue<'lua>,
        lua: &'lua mlua::prelude::Lua,
    ) -> mlua::prelude::LuaResult<Self> {
        Ok(Self::new(String::from_lua(value, lua)?))
    }
}

impl Serialize for PackageName {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        self.name.serialize(serializer)
    }
}

impl From<&str> for PackageName {
    fn from(value: &str) -> Self {
        Self::new(value.into())
    }
}

impl Display for PackageName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.name.as_str())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn parse_name() {
        let mut package_name: PackageName = "neorg".into();
        assert_eq!(package_name.to_string(), "neorg");
        package_name = "LuaFileSystem".into();
        assert_eq!(package_name.to_string(), "luafilesystem");
    }

    #[tokio::test]
    async fn parse_lua_package() {
        let neorg = LuaPackage::parse("neorg".into(), "1.0.0".into()).unwrap();
        let expected_version = PackageVersion::parse("1.0.0").unwrap();
        assert_eq!(neorg.name().to_string(), "neorg");
        assert_eq!(*neorg.version(), expected_version);
        let neorg = LuaPackage::parse("neorg".into(), "1.0".into()).unwrap();
        assert_eq!(*neorg.version(), expected_version);
        let neorg = LuaPackage::parse("neorg".into(), "1".into()).unwrap();
        assert_eq!(*neorg.version(), expected_version);
    }

    #[tokio::test]
    async fn parse_lua_package_req() {
        let mut package_req = LuaPackageReq::new("foo".into(), Some("1.0.0".into())).unwrap();
        assert!(package_req.matches(&LuaPackage::parse("foo".into(), "1.0.0".into()).unwrap()));
        assert!(!package_req.matches(&LuaPackage::parse("bar".into(), "1.0.0".into()).unwrap()));
        assert!(!package_req.matches(&LuaPackage::parse("foo".into(), "2.0.0".into()).unwrap()));
        package_req = LuaPackageReq::new("foo".into(), Some(">= 1.0.0".into())).unwrap();
        assert!(package_req.matches(&LuaPackage::parse("foo".into(), "2.0.0".into()).unwrap()));
        let package_req: LuaPackageReq = "lua >= 5.1".parse().unwrap();
        assert_eq!(package_req.name.to_string(), "lua");
        let package_req: LuaPackageReq = "lua>=5.1".parse().unwrap();
        assert_eq!(package_req.name.to_string(), "lua");
        let package_req: LuaPackageReq = "toml-edit >= 0.1.0".parse().unwrap();
        assert_eq!(package_req.name.to_string(), "toml-edit");
        let package_req: LuaPackageReq = "plugin.nvim >= 0.1.0".parse().unwrap();
        assert_eq!(package_req.name.to_string(), "plugin.nvim");
        let package_req: LuaPackageReq = "lfs".parse().unwrap();
        assert_eq!(package_req.name.to_string(), "lfs");
        let package_req: LuaPackageReq = "neorg 1.0.0".parse().unwrap();
        assert_eq!(package_req.name.to_string(), "neorg");
        let neorg = LuaPackage::parse("neorg".into(), "1.0.0".into()).unwrap();
        assert!(package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "2.0.0".into()).unwrap();
        assert!(!package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg 2.0.0".parse().unwrap();
        assert!(package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg = 2.0.0".parse().unwrap();
        assert!(package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg == 2.0.0".parse().unwrap();
        assert!(package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg &equals; 2.0.0".parse().unwrap();
        assert!(package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg >= 1.0, &lt; 2.0".parse().unwrap();
        let neorg = LuaPackage::parse("neorg".into(), "1.5".into()).unwrap();
        assert!(package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg &gt; 1.0, &lt; 2.0".parse().unwrap();
        let neorg = LuaPackage::parse("neorg".into(), "1.11.0".into()).unwrap();
        assert!(package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "3.0.0".into()).unwrap();
        assert!(!package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "0.5".into()).unwrap();
        assert!(!package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg ~> 1".parse().unwrap();
        assert!(!package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "3".into()).unwrap();
        assert!(!package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "1.5".into()).unwrap();
        assert!(package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg ~> 1.4".parse().unwrap();
        let neorg = LuaPackage::parse("neorg".into(), "1.3".into()).unwrap();
        assert!(!package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "1.5".into()).unwrap();
        assert!(!package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "1.4.10".into()).unwrap();
        assert!(package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "1.4".into()).unwrap();
        assert!(package_req.matches(&neorg));
        let package_req: LuaPackageReq = "neorg ~> 1.0.5".parse().unwrap();
        let neorg = LuaPackage::parse("neorg".into(), "1.0.4".into()).unwrap();
        assert!(!package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "1.0.5".into()).unwrap();
        assert!(package_req.matches(&neorg));
        let neorg = LuaPackage::parse("neorg".into(), "1.0.6".into()).unwrap();
        assert!(!package_req.matches(&neorg));
    }
}
