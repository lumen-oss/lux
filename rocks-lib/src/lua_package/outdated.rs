use eyre::{eyre, Result};

use crate::manifest::ManifestMetadata;

use super::{version::PackageVersion, LuaPackage};

impl LuaPackage {
    /// Tries to find a newer version of a given rock.
    /// Returns the latest version if found.
    pub fn has_update(&self, manifest: &ManifestMetadata) -> Result<Option<PackageVersion>> {
        let latest_version: PackageVersion = manifest
            .latest_version(&self.name)
            .ok_or(eyre!("rock {} not found!", self.name))?
            .parse()?;

        if self.version < latest_version {
            Ok(Some(latest_version))
        } else {
            Ok(None)
        }
    }
}
