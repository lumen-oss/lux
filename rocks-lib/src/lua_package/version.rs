use std::{cmp::Ordering, fmt::Display, str::FromStr};

use eyre::Result;
use html_escape::decode_html_entities;
use mlua::FromLua;
use semver::{Comparator, Error, Op, Version, VersionReq};
use serde::{de, Deserialize, Deserializer, Serialize};

/// **SemVer version** as defined by <https://semver.org>.
/// or a **Dev** version, which can be one of "dev", "scm", or "git"
#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub enum PackageVersion {
    SemVer { version: Version },
    Dev { name: String },
}

impl PackageVersion {
    pub fn parse(text: &str) -> Result<Self> {
        PackageVersion::from_str(text)
    }
}

impl Serialize for PackageVersion {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        match self {
            PackageVersion::SemVer { version } => version.to_string().serialize(serializer),
            PackageVersion::Dev { name } => name.serialize(serializer),
        }
    }
}

impl<'de> Deserialize<'de> for PackageVersion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        Self::from_str(&s).map_err(de::Error::custom)
    }
}

impl<'lua> FromLua<'lua> for PackageVersion {
    fn from_lua(
        value: mlua::prelude::LuaValue<'lua>,
        lua: &'lua mlua::prelude::Lua,
    ) -> mlua::prelude::LuaResult<Self> {
        let s = String::from_lua(value, lua)?;
        Self::from_str(&s).map_err(|err| mlua::Error::DeserializeError(err.to_string()))
    }
}

impl Display for PackageVersion {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PackageVersion::SemVer { version } => version.fmt(f),
            PackageVersion::Dev { name } => f.write_str(name.as_str()),
        }
    }
}

impl PartialOrd for PackageVersion {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for PackageVersion {
    fn cmp(&self, other: &Self) -> Ordering {
        match (self, other) {
            (PackageVersion::SemVer { version: a }, PackageVersion::SemVer { version: b }) => {
                a.cmp(b)
            }
            (PackageVersion::SemVer { .. }, PackageVersion::Dev { .. }) => Ordering::Less,
            (PackageVersion::Dev { .. }, PackageVersion::SemVer { .. }) => Ordering::Greater,
            (PackageVersion::Dev { name: a }, PackageVersion::Dev { name: b }) => a.cmp(b),
        }
    }
}

impl FromStr for PackageVersion {
    type Err = eyre::Error;

    fn from_str(text: &str) -> Result<Self, Self::Err> {
        if is_dev_version_str(text) {
            return Ok(PackageVersion::Dev { name: text.into() });
        }
        Ok(PackageVersion::SemVer {
            version: parse_version(text)?,
        })
    }
}

/// **SemVer version** requirement as defined by <https://semver.org>.
/// or a **Dev** version requirement, which can be one of "dev", "scm", or "git"
#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub enum PackageVersionReq {
    SemVer { version_req: VersionReq },
    Dev { name_req: String },
}

impl PackageVersionReq {
    pub fn parse(text: &str) -> Result<Self> {
        PackageVersionReq::from_str(text)
    }
    pub fn matches(&self, version: &PackageVersion) -> bool {
        match (self, version) {
            (PackageVersionReq::SemVer { version_req }, PackageVersion::SemVer { version }) => {
                version_req.matches(version)
            }
            (PackageVersionReq::SemVer { .. }, PackageVersion::Dev { .. }) => true,
            (PackageVersionReq::Dev { .. }, PackageVersion::SemVer { .. }) => false,
            (PackageVersionReq::Dev { name_req }, PackageVersion::Dev { name }) => {
                name_req.ends_with(name)
            }
        }
    }
}

impl Display for PackageVersionReq {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PackageVersionReq::SemVer { version_req } => version_req.fmt(f),
            PackageVersionReq::Dev { name_req } => f.write_str(name_req.as_str()),
        }
    }
}

impl Default for PackageVersionReq {
    fn default() -> Self {
        PackageVersionReq::SemVer {
            version_req: VersionReq::default(),
        }
    }
}

impl From<PackageVersion> for PackageVersionReq {
    fn from(version: PackageVersion) -> Self {
        match version {
            PackageVersion::Dev { name } => PackageVersionReq::Dev { name_req: name },
            PackageVersion::SemVer { version } => PackageVersionReq::SemVer {
                version_req: VersionReq {
                    comparators: vec![Comparator {
                        op: Op::Exact,
                        major: version.major,
                        minor: Some(version.minor),
                        patch: Some(version.patch),
                        pre: version.pre,
                    }],
                },
            },
        }
    }
}

impl FromStr for PackageVersionReq {
    type Err = eyre::Error;

    fn from_str(text: &str) -> Result<Self, Self::Err> {
        if is_dev_version_str(text.trim_start_matches("==").trim()) {
            return Ok(PackageVersionReq::Dev {
                name_req: text.into(),
            });
        }
        Ok(PackageVersionReq::SemVer {
            version_req: parse_version_req(text)?,
        })
    }
}

fn is_dev_version_str(text: &str) -> bool {
    matches!(text, "dev" | "scm" | "git")
}

/// Parses a Version from a string, automatically supplying any missing details (i.e. missing
/// minor/patch sections).
fn parse_version(s: &str) -> Result<Version, Error> {
    Version::parse(&append_minor_patch_if_missing(s.to_string()))
}

/// Transform LuaRocks constraints into constraints that can be parsed by the semver crate.
fn parse_version_req(version_constraints: &str) -> Result<VersionReq, Error> {
    let unescaped = decode_html_entities(version_constraints)
        .to_string()
        .as_str()
        .to_owned();
    let transformed = match unescaped {
        s if s.starts_with("~>") => parse_pessimistic_version_constraint(s)?,
        // The semver crate only understands "= version", unlike luarocks which understands "== version".
        s if s.starts_with("==") => s[1..].to_string(),
        s => s,
    };

    let version_req = VersionReq::parse(&transformed)?;
    Ok(version_req)
}

fn parse_pessimistic_version_constraint(version_constraint: String) -> Result<String, Error> {
    // pessimistic operator
    let min_version_str = &version_constraint[2..].trim();
    let min_version = Version::parse(&append_minor_patch_if_missing(min_version_str.to_string()))?;

    let max_version = match min_version_str.matches('.').count() {
        0 => Version {
            major: &min_version.major + 1,
            ..min_version.clone()
        },
        1 => Version {
            minor: &min_version.minor + 1,
            ..min_version.clone()
        },
        _ => Version {
            patch: &min_version.patch + 1,
            ..min_version.clone()
        },
    };

    Ok(format!(">= {min_version}, < {max_version}"))
}

/// Recursively append .0 until the version string has a minor or patch version
fn append_minor_patch_if_missing(version: String) -> String {
    if version.matches('.').count() < 2 {
        append_minor_patch_if_missing(version + ".0")
    } else {
        version
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn parse_dev_version() {
        assert_eq!(
            PackageVersion::parse("dev").unwrap(),
            PackageVersion::Dev { name: "dev".into() }
        );
        assert_eq!(
            PackageVersion::parse("scm").unwrap(),
            PackageVersion::Dev { name: "scm".into() }
        );
        assert_eq!(
            PackageVersion::parse("git").unwrap(),
            PackageVersion::Dev { name: "git".into() }
        );
    }

    #[tokio::test]
    async fn parse_dev_version_req() {
        assert_eq!(
            PackageVersionReq::parse("dev").unwrap(),
            PackageVersionReq::Dev {
                name_req: "dev".into()
            }
        );
        assert_eq!(
            PackageVersionReq::parse("scm").unwrap(),
            PackageVersionReq::Dev {
                name_req: "scm".into()
            }
        );
        assert_eq!(
            PackageVersionReq::parse("git").unwrap(),
            PackageVersionReq::Dev {
                name_req: "git".into()
            }
        );
        assert_eq!(
            PackageVersionReq::parse("==dev").unwrap(),
            PackageVersionReq::Dev {
                name_req: "==dev".into()
            }
        );
        assert_eq!(
            PackageVersionReq::parse("==git").unwrap(),
            PackageVersionReq::Dev {
                name_req: "==git".into()
            }
        );
        assert_eq!(
            PackageVersionReq::parse("== dev").unwrap(),
            PackageVersionReq::Dev {
                name_req: "== dev".into()
            }
        );
        assert_eq!(
            PackageVersionReq::parse("== scm").unwrap(),
            PackageVersionReq::Dev {
                name_req: "== scm".into()
            }
        );
    }
}
