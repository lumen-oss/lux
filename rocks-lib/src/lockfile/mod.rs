use std::io::Write;
use std::{collections::HashMap, fs::File, io::ErrorKind, path::PathBuf};

use eyre::{bail, Result};
use itertools::Itertools;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::lua_package::{LuaPackage, PackageName, PackageVersion};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct LockedPackage {
    pub name: PackageName,
    pub version: PackageVersion,
    pub pinned: bool,
    pub dependencies: Vec<String>,
}

impl LockedPackage {
    pub fn from(package: &LuaPackage) -> Self {
        Self {
            name: package.name().clone(),
            version: package.version().clone(),
            pinned: false,
            dependencies: Vec::default(),
        }
    }

    pub fn pinned(mut self, pin: bool) -> Self {
        self.pinned = pin;
        self
    }

    pub fn id(&self) -> String {
        let mut hasher = Sha256::new();

        hasher.update(format!("{}{}{}", self.name, self.version, self.pinned));

        hex::encode(hasher.finalize())
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Lockfile {
    #[serde(skip_serializing, skip_deserializing)]
    filepath: PathBuf,
    // TODO: Serialize this directly into a `Version`
    version: String,
    // NOTE: We cannot directly serialize to a `Sha256` object as they don't implement serde traits.
    rocks: HashMap<String, LockedPackage>,
    entrypoints: Vec<String>,
}

impl Lockfile {
    pub fn new(filepath: PathBuf) -> Result<Self> {
        // Ensure that the lockfile exists
        match File::options().create_new(true).write(true).open(&filepath) {
            Ok(mut file) => {
                write!(
                    file,
                    r#"
                        {{
                            "entrypoints": [],
                            "rocks": {{}},
                            "version": "1.0.0"
                        }}
                    "#
                )?;
            }
            Err(err) if err.kind() == ErrorKind::AlreadyExists => {}
            Err(err) => bail!(err),
        }

        let mut new: Lockfile = serde_json::from_str(&std::fs::read_to_string(&filepath)?)?;

        new.filepath = filepath;

        Ok(new)
    }

    pub fn add(&mut self, rock: &LuaPackage, pinned: bool) -> LockedPackage {
        let rock = LockedPackage::from(rock).pinned(pinned);

        self.rocks.entry(rock.id()).or_insert(rock).clone()
    }

    pub fn add_dependency(&mut self, target: &LockedPackage, dependency: &LockedPackage) {
        let target_id = target.id();
        let dependency_id = dependency.id();

        self.rocks
            .entry(target_id)
            .and_modify(|rock| rock.dependencies.push(dependency_id));
    }

    pub fn remove(&mut self, target: &LockedPackage) {
        self.rocks.remove(&target.id());
    }

    pub fn version(&self) -> &String {
        &self.version
    }

    pub fn rocks(&self) -> &HashMap<String, LockedPackage> {
        &self.rocks
    }

    // TODO: `fn entrypoints() -> Vec<LockedRock>`

    pub fn flush(&mut self) -> Result<()> {
        let dependencies = self
            .rocks
            .iter()
            .flat_map(|(_, rock)| &rock.dependencies)
            .collect_vec();

        self.entrypoints = self
            .rocks
            .keys()
            .filter(|id| !dependencies.iter().contains(id))
            .cloned()
            .collect();

        let content = serde_json::to_string(self)?;

        std::fs::write(&self.filepath, content)?;

        Ok(())
    }
}

impl Drop for Lockfile {
    fn drop(&mut self) {
        let _ = self.flush();
    }
}

#[cfg(test)]
mod tests {
    use std::{fs::remove_file, path::PathBuf};

    use assert_fs::fixture::PathCopy;
    use insta::{assert_json_snapshot, sorted_redaction};

    use crate::{config::LuaVersion::Lua51, lua_package::LuaPackage, tree::Tree};

    #[test]
    fn parse_lockfile() {
        let temp = assert_fs::TempDir::new().unwrap();
        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("resources/test/sample-tree"),
            &["**"],
        )
        .unwrap();

        let tree = Tree::new(temp.to_path_buf(), Lua51).unwrap();
        let lockfile = tree.lockfile().unwrap();

        assert_json_snapshot!(lockfile, { ".**" => sorted_redaction() });
    }

    #[test]
    fn add_rocks() {
        let temp = assert_fs::TempDir::new().unwrap();
        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("resources/test/sample-tree"),
            &["**"],
        )
        .unwrap();

        let tree = Tree::new(temp.to_path_buf(), Lua51).unwrap();
        let mut lockfile = tree.lockfile().unwrap();

        let test_package = LuaPackage::parse("test1".to_string(), "0.1.0".to_string()).unwrap();
        let package = lockfile.add(&test_package, false);

        let test_package = LuaPackage::parse("test2".to_string(), "0.1.0".to_string()).unwrap();
        let dependency = lockfile.add(&test_package, true);

        lockfile.add_dependency(&package, &dependency);

        assert_json_snapshot!(lockfile, { ".**" => sorted_redaction() });
    }

    #[test]
    fn parse_nonexistent_lockfile() {
        let tree_path =
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("resources/test/sample-tree");

        let temp = assert_fs::TempDir::new().unwrap();
        temp.copy_from(&tree_path, &["**"]).unwrap();

        remove_file(temp.join("5.1/lock.json")).unwrap();

        let tree = Tree::new(temp.to_path_buf(), Lua51).unwrap();

        tree.lockfile().unwrap(); // Try to create the lockfile but don't actually do anything with it.
    }
}
