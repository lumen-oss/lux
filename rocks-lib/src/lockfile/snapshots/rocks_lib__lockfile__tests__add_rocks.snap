---
source: rocks-lib/src/lockfile/mod.rs
expression: lockfile
---
{
  "version": "1.0.0",
  "rocks": {
    "48ec344951668eca0e0a4ff284d804a11e4e709194df3191a72ed8fac89cf2e0": {
      "name": "lua-cjson",
      "version": "2.1.0",
      "pinned": false,
      "dependencies": []
    },
    "aa0a5bcd396f3b8e59fa9dd8059a06c3bf4952f48473214d96fc46895edb59f0": {
      "name": "neorg",
      "version": "8.8.1",
      "pinned": false,
      "dependencies": [
        "48ec344951668eca0e0a4ff284d804a11e4e709194df3191a72ed8fac89cf2e0"
      ]
    },
    "ae010d6b758c43148b300c47dbe56532df65e0b4151759ba8bea7e22e787939c": {
      "name": "test2",
      "version": "0.1.0",
      "pinned": true,
      "dependencies": []
    },
    "d27609febb7bc0a8b6be63fb51b8a2daed3dd27e03badaa2ebfbbc73ac658f8b": {
      "name": "test1",
      "version": "0.1.0",
      "pinned": false,
      "dependencies": [
        "ae010d6b758c43148b300c47dbe56532df65e0b4151759ba8bea7e22e787939c"
      ]
    },
    "d7164c4921869877c7f29fce3f14b8ea78e0aec0d0c36123a3a920a894785ea5": {
      "name": "neorg",
      "version": "8.0.0",
      "pinned": false,
      "dependencies": [
        "48ec344951668eca0e0a4ff284d804a11e4e709194df3191a72ed8fac89cf2e0"
      ]
    }
  },
  "entrypoints": [
    "aa0a5bcd396f3b8e59fa9dd8059a06c3bf4952f48473214d96fc46895edb59f0",
    "d7164c4921869877c7f29fce3f14b8ea78e0aec0d0c36123a3a920a894785ea5"
  ]
}
